bentoml/_internal/server/http_app.py

```diff
@@ -264,6 +264,15 @@ def on_startup(self) -> list[t.Callable[[], None]]:
         else:
             for runner in self.bento_service.runners:
                 on_startup.append(runner.init_client)
+
+        async def wait_for_runner_ready():
+            ready_status = False
+            while not ready_status:
+                ready_status = all(await asyncio.gather(*(runner.runner_handle_is_ready() for runner in self.bento_service.runners)))
+                await asyncio.sleep(0.1)
+        
+        on_startup.append(lambda: asyncio.run(wait_for_runner_ready()))
```

Comment:
This won't work because we are already inside a running event loop
```suggestion
        on_startup.append(
            lambda: asyncio.ensure_future(
                wait_runner_ready(), loop=asyncio.get_running_loop()
            )
        )
```
I don't think you need to sleep on L272 either.

Reply:
Thanks. I implemented the suggested change and got the below message from VS Code. I'm looking into this but wanted to share in case you have any thoughts.
```
Argument of type "() -> Task[None]" cannot be assigned to parameter "__object" of type "() -> None" in function "append"
  Type "() -> Task[None]" cannot be assigned to type "() -> None"
    Function return type "Task[None]" is incompatible with type "None"
```

I'll remove sleep on line 272.

Reply:
We should probably sleep so as not to send a bunch of spam packets until the runner is ready; honestly I'd think about increasing it to something like 5s to mirror the kubernetes ready check intervals.

Reply:
Re: `ensure_future`: it should be `get_event_loop().run_until_complete(...)`; I was wrong about the deprecation, that's just outside an event loop.

Reply:
`run_until_complete` won't work since the loop is already runner.

Reply:
Oh, our type annotations are just wrong and this should work fine:
```suggestion
        on_startup.append(wait_for_runner_ready)
```

We should also extend the `on_startup` types to allow coroutines as well as functions.

Is there actually a way to run an async function inside of a non-async function inside a running event loop?

Also, why is it that we aren't failing e2e tests?

Reply:
you have to call `run_sync`

Reply:
re for the question, i don't believe there is such thing

Reply:
@sauyon
1. Could you give me pointer on how to properly extend `on_startup` to return multiple types? It looks like [Union](https://docs.python.org/3/library/typing.html#typing.Union) is the solution. I imported `Union` and tried `def on_startup(self) -> Union[list[t.Callable[[], None]], Coroutine]` but got the following message:
```Declared return type, "list[() -> None] | Coroutine", is partially unknown```.

2. It looks like after we do the above, we also need to do the same for `on_startup` in `base_app.py`.

3. Side question on how typing is implemented in the code base: In ` -> list[t.Callable[[], None]]`, I understand that `t.Callable[[], None]` allows us to return a function, but what does `list[]` do? Does it allow us to return a list of functions?

@aarnphm 
Confirming that you were referring to `run_sync` from [Tornado](https://www.tornadoweb.org/en/stable/ioloop.html)? If so, should we be concerned about this docs' statement that "The IOLoop interface is now provided primarily for backwards compatibility; new code should generally use the asyncio event loop interface directly."?

Would the proper implementation of `run_sync` look something like this?
```
import tornado.ioloop

on_startup.append(ioloop.IOLoop.run_sync(wait_for_runner_ready))
```